### PR TITLE
FileRef: Allow an IOStream to be used

### PR DESCRIPTION
**Pull request by chouquette**
Refs issue #526 

I haven't tested it on WinRT (which is the main reason for this, but it works fine on linux)
I'm submitting this for early review and will post the WinRT results as soon as I have a build & it's integrated.

Cheers,

### CODE REVIEW

File: taglib/fileref.cpp [hunk 1]
```diff
@@ -284,3 +289,77 @@ File *FileRef::create(FileName fileName, bool readAudioProperties,
...
+      const int pos = s.rfind(".");
+      if(pos != -1)
+        ext = s.substr(pos + 1).upper();
+    }
+
+    // If this list is updated, the method defaultFileExtensions() should also be
+    // updated.  However at some point that list should be created at the same time
+    // that a default file type resolver is created.
+
+    if(!ext.isEmpty()) {
```
Comment (lalinsky) [inline taglib/fileref.cpp]: Please do not make a copy of this whole block. Create an internal template function instead.

**Comment by chouquette** [comment]
Updated with a template function to avoid duplicating code

File: taglib/fileref.cpp [hunk 2]
```diff
@@ -54,6 +54,62 @@
 
 using namespace TagLib;
 
+template <typename T>
+static File* createInternal(T arg, const String& ext, bool readAudioProperties,
```
Comment (lalinsky) [inline taglib/fileref.cpp]: The extension extraction code could be here as well. Just pass it `FileName fileName` instead of `String ext`.

**Comment by lalinsky** [comment]
Can you please add at least one test of this new API to `tests/test_fileref.cpp`?

File: taglib/fileref.cpp (hunk 2 above)
Comment (chouquette) [inline taglib/fileref.cpp]: I thought of that, but it makes me pass filename twice for one version, and iostream, filename for the other... That feels weird, so I refrained myself.
There's a way to guess the filename based on the type that is provided, but I'm not sure that would improve readability...
If you're fine with passing filename twice I'll add it

Comment (lalinsky) [inline taglib/fileref.cpp]: I'm fine with passing the filename twice, of course. That's why I suggested it. :)

Comment (chouquette) [inline taglib/fileref.cpp]: Just wanted to check we had the same understanding :)
I just pushed that change as well, and updated the tests so all tests are also using the IOStream constructor.

**Comment by chouquette** [comment]
I forgot to mention it before, but this works fine for VLC on WinRT :)

File: taglib/fileref.cpp [hunk 3]
```diff
@@ -213,74 +278,22 @@ File *FileRef::create(FileName fileName, bool readAudioProperties,
...
-    if(ext == "XM")
-      return new XM::File(fileName, readAudioProperties, audioPropertiesStyle);
-  }
-
-  return 0;
+File *FileRef::create(IOStream* ioStream, bool readAudioProperties, AudioProperties::ReadStyle audioPropertiesStyle)
+{
+  return createInternal(ioStream,
+#ifdef _WIN32
+  ioStream->name().toString();
```
Comment (scotchi) [inline taglib/fileref.cpp]: This almost certainly is wrong.  `;` without a closing paren?

File: taglib/fileref.cpp [hunk 4]
```diff
@@ -54,6 +54,67 @@
 
 using namespace TagLib;
 
+template <typename T>
+static File* createInternal(T arg, const String& filename, bool readAudioProperties,
```
Comment (scotchi) [inline taglib/fileref.cpp]: TagLib aligns `*` and `&` to the right, so `static File *create...`

File: taglib/fileref.cpp [hunk 5]
```diff
@@ -54,6 +54,67 @@
...
+                            AudioProperties::ReadStyle audioPropertiesStyle)
+{
+    String ext;
```
Comment (scotchi) [inline taglib/fileref.cpp]: Please use 2 space indents.

**Comment by TsudaKageyu** [comment]
@chouquette, sorry for making you wait for a long time. This PR has been merged now.

**Pull request merged**